transport/internal/tlsmux/mux.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tlsmux
+
+import (
+	"crypto/tls"
+	"net"
+)
+
+// mux provides a listener which accepts both TLS and non-TLS connections.
+type mux struct {
+	net.Listener
+	tlsConfig *tls.Config
+}
+
+// NewListener returns a multiplexed listener which accepts both TLS and non-TLS connections.
+func NewListener(listener net.Listener, tlsConfig *tls.Config) net.Listener {
+	return &mux{
+		Listener:  listener,
+		tlsConfig: tlsConfig,
+	}
+}
+
+// Accept returns the multiplexed connetions.
+func (m *mux) Accept() (net.Conn, error) {
+	conn, err := m.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(jronak): avoid slow connections causing head of the line blocking by spawning
+	// connection processing in separate routine.
+
+	return m.mux(conn)
+}
+
+func (m *mux) mux(conn net.Conn) (net.Conn, error) {
```

Comment:
nit: this function name is quite confusing since struct name is also mux. This function returns the actual connection object, should be keep the name something more relevant?

Reply:
Makes sense, renamed it to `handle`